### PR TITLE
Add support for a single player

### PR DESCRIPTION
**Pull request by jonathan-marmelab**
This PR closes the following user story : https://trello.com/c/5NHQCXa5/1-3-as-robert-i-can-play-my-turn

- [X] Player support
- [X] Move pawn
- [X] Move pawn with tile if necessary 

Capture d’écran du 2023-02-22 15-24-38
Capture d’écran du 2023-02-22 15-28-47

### CODE REVIEW

File: internal/model/board.go [hunk 1]
```diff
@@ -204,14 +295,59 @@ func NewBoard(size int) (*Board, error) {
...
 	}
 
-	for i := 0; i < size; i++ {
-		board.Tiles[i] = make([]*BoardTile, size)
-		for j := 0; j < size; j++ {
-			board.Tiles[i][j] = &BoardTile{
-				Tile:     tiles[(size*i)+j],
-				Rotation: randomRotation(),
+	tileIndex := 0
+	for line := 0; line < size; line++ {
```
Comment (arimet) [inline internal/model/board.go:309]: Maybe a comment to explain why you do this

Comment (jonathan-marmelab) [reply internal/model/board.go:314]: Done

File: internal/model/board.go [hunk 2]
```diff
@@ -27,61 +37,114 @@ type Board struct {
...
 	RemainingTile *BoardTile `json:"remainingTile"`
+
+	// Players holds the players that are part of the game.
+	Players []*Player `json:"players"`
+
+	// GameState is the current game state
+	State GameState `json:"gameState"`
+}
+
+func (b Board) tryPlaceTile(index int) error {
```
Comment (fzaninotto) [inline internal/model/board.go:48]: Why a lowercase "T" while other function names start with a capital letter?

Also, the word "try" is weird. I see this function as a validator, so why not call it `validateTileMove` instead?

File: internal/model/board.go [hunk 3]
```diff
@@ -139,17 +227,23 @@ const (
...
 
+// generate tiles generates tile list for the given board size.
+// It will only generate size*size - 3 cards, since the tiles on each corner is
+// predefined (V shape).
 func generateTiles(size int) (tiles []*Tile, treasureCount int) {
 	var (
-		tileCount                = size*size + 1
+		tileCount = size*size + 1
+
+		// We need to generate 4 less tiles as the corners are V tiles
```
Comment (fzaninotto) [inline internal/model/board.go:237]:
```suggestion
		// We need to generate 4 less tiles as the corners are fixed V tiles
```

File: internal/presentation/gui.go [hunk 4]
```diff
@@ -14,26 +14,75 @@ type gameUi struct {
...
+func (g gameUi) foregroundColor() gocui.Attribute {
+	switch g.board.CurrentPlayer().Color {
+	case model.ColorBlue, model.ColorGreen, model.ColorRed:
+		return gocui.ColorWhite
+	default:
+		return gocui.ColorBlack
+	}
+}
+
+func (g gameUi) drawButton(name, text string, topLeftX, topLeftY, bottomRightX, bottomRightY int, handler GuiHandler) (*gocui.View, error) {
```
Comment (fzaninotto) [inline internal/presentation/gui.go:42]: I thought the practice in go is to return the error first. Out of curiosity, why do you return the gocuiView first here?

File: internal/presentation/loop.go [hunk 5]
```diff
@@ -33,29 +33,33 @@ type GuiHandler func(gui *gocui.Gui, view *gocui.View) error
...
+	board     *model.Board
 	saveBoard storage.BoardSaveFn
 
 	gui *gocui.Gui
 }
 
 func (g gameLoop) insertTile(button Button, buttonIndex int) GuiHandler {
 	return func(gui *gocui.Gui, view *gocui.View) error {
+		if g.board.State != model.GameStatePlaceTile {
+			return nil
```
Comment (fzaninotto) [inline internal/presentation/loop.go:45]: shouldn't you return an error instead?

File: internal/presentation/loop.go [hunk 6]
```diff
@@ -68,9 +72,20 @@ func (g gameLoop) rotateRemainingTile(rotationType RotationType) GuiHandler {
...
+			g.board.RotateRemainingTileAntiClockwise()
+		}
+		return g.saveBoard()
+	}
+}
+
+func (g gameLoop) moveCurrentPlayerTo(line, row int) GuiHandler {
+	return func(gui *gocui.Gui, view *gocui.View) error {
+		if err := g.board.MoveCurrentPlayerTo(line, row); err != nil {
+			if err != model.ErrInvalidAction {
```
Comment (fzaninotto) [inline internal/presentation/loop.go:86]: I don't understand the rationale of this test. Why should you swallow invalid action errors?

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Excellent! Just a few minor remarks.

File: internal/presentation/loop.go (hunk 5 above)
Comment (jonathan-marmelab) [reply internal/presentation/loop.go:45]: If I return an error here, the program will crash as any error returned in an handler function makes the program stop.

File: internal/presentation/loop.go (hunk 6 above)
Comment (jonathan-marmelab) [reply internal/presentation/loop.go:86]: This is the same reason as previous comment, I will add a comment to provide the reason in the code.

File: internal/presentation/loop.go (hunk 5 above)
Comment (jonathan-marmelab) [reply internal/presentation/loop.go:45]: I will add a comment to provide the reason in the code.

File: internal/presentation/gui.go (hunk 4 above)
Comment (jonathan-marmelab) [reply internal/presentation/gui.go:42]: The error is the last parameter in idiomatic Go (See https://go.dev/blog/error-handling-and-go).

File: internal/model/board.go (hunk 2 above)
Comment (jonathan-marmelab) [reply internal/model/board.go:48]: The lower-case letter is used to define private methods.

I initialy used `try` to make a parallel with exception handling, but `validate` is definitly clearer. Thank you !

**Review by arimet — APPROVED** [review]

**Pull request merged**